### PR TITLE
Initial base fee support

### PR DESCRIPTION
**Pull request by forshtat**
More base fee support is on its way

### CODE REVIEW

File: src/js/relayclient/ServerHelper.js [hunk 1]
```diff
@@ -143,7 +138,7 @@ class ServerHelper {
   defaultCalculateRelayScore (relay) {
     // basic score is trasnaction fee (which is %)
     // higher the better.
-    let score = 1000 - relay.transactionFee
+    let score = 1000 - relay.pctRelayFee
```
Comment (drortirosh) [inline src/js/relayclient/ServerHelper.js]: this default score is broken: 
I can put `baseFee = 1e18, pctRelayFee=0`, and I'll always be ordered as first...
The real calculation should estimate the (inner) tx gas price, and use the `base+estimate*pct` for score.

File: src/js/relayclient/RelayClient.js [hunk 2]
```diff
@@ -495,11 +511,11 @@ class RelayClient {
           console.log('relayTransaction: req:', {
             from: options.from,
             to: options.to,
-            encodedFunctionCall,
-            txfee: options.txfee,
+            encodedFunctionCall: encodedFunction,
+            pctRelayFee: options.pctRelayFee,
```
Comment (drortirosh) [inline src/js/relayclient/RelayClient.js]: baseFee?

File: .gitignore [hunk 3]
```diff
@@ -19,3 +19,4 @@
 /server/pkg/
 /.eslintcache
 /src/js/relayclient/interfaces/
+/ganache.pid
```
Comment (drortirosh) [inline .gitignore]: who creates it? why here?

File: contracts/samples/KnownUsersSponsor.sol [hunk 4]
```diff
@@ -1,82 +0,0 @@
-pragma solidity ^0.5.16;
```
Comment (drortirosh) [inline contracts/samples/KnownUsersSponsor.sol]: why did you remove it/

File: server/src/librelay/relay_server.go [hunk 5]
```diff
@@ -668,7 +681,7 @@ func (relay *RelayServer) canRelay(from common.Address,
 }
 
 func (relay *RelayServer) validateFee(relayFee big.Int) bool {
```
Comment (drortirosh) [inline server/src/librelay/relay_server.go]: wrong test: should be `pctFee>0 || baseFee > 0`

File: server/src/librelay/relay_server_test.go [hunk 6]
```diff
@@ -407,7 +408,7 @@ func assertNoTransactionResent(t *testing.T, relay *RelayServer) {
 }
 
 func TestCreateRelayTransaction(t *testing.T) {
-	request := newRelayTransactionRequest(t, 0, "0x717591514ee3e28f3981d9b1a9a80a4d5b8a2a824ad6d4329a83272e14ecf87b77067fa322eb5142d93bedd488d1a34aea2faef8c288cbf4603e55211f34e5171c")
+	request := newRelayTransactionRequest(t, 0, "0xc2f3ccc4a624ca99cbd7928503f33347a3dcaaf624d8640610c0193e7b3ae868409d84ad4a97c8f01c0bd8f9b65bd81a9efd44b865ad75df13bdfb512d809b911b")
```
Comment (drortirosh) [inline server/src/librelay/relay_server_test.go]: these opaque strings are still here, with no explanation on how to generate them.

File: src/js/relayclient/RelayClient.js [hunk 7]
```diff
@@ -441,23 +459,20 @@ class RelayClient {
...
 
-        const _gasLimits = await gasSponsorContract.methods.getGasLimitsForSponsorCalls().call()
-        const _overhead = parseInt(await relayHub.methods.getHubOverhead().call())
-        const gasData = await utils.getTransactionGasData({
-          _gasLimits,
-          _overhead,
-          relayHub,
-          calldataSize,
-          gtxdatanonzero: options.gtxdatanonzero || Environments.default.gtxdatanonzero,
+        const gasLimits = await paymasterContract.methods.getGasLimits().call()
```
Comment (drortirosh) [inline src/js/relayclient/RelayClient.js]: this is far too much work to do before `canRelay`, just like you don't do all these calculations before calling `relayCall`, you shouldn't here:
need to have a wrapper for canRelay which get all the needed vallues (hub overhead, gasLimits etc))

File: src/js/relayclient/RelayClient.js [hunk 8]
```diff
@@ -536,10 +552,10 @@ class RelayClient {
     const params = payload.params[0]
     const relayClientOptions = this.config
 
-    const { txfee, txFee, gas, gasPrice } = params
+    const { pctRelayFee, txFee, gas, gasPrice } = params
     const relayOptions = {
       ...params,
-      txfee: txFee || txfee || relayClientOptions.txfee,
+      pctRelayFee: txFee || pctRelayFee || relayClientOptions.pctRelayFee,
```
Comment (drortirosh) [inline src/js/relayclient/RelayClient.js]: missing baseFee

File: src/js/relayclient/ServerHelper.js (hunk 1 above)
Comment (drortirosh) [reply src/js/relayclient/ServerHelper.js]: also, misisng tests for various baseFee values

File: server/src/librelay/relay_server.go (hunk 5 above)
Comment (forshtat) [reply server/src/librelay/relay_server.go]: go server code is nearing end-of-life if we stick to the plan. Did write down the task to check JS server supports baseFee correctly

File: server/src/librelay/relay_server_test.go (hunk 6 above)
Comment (forshtat) [reply server/src/librelay/relay_server_test.go]: There is an explanation on how to generate them printed to console if test fails. again, this go code is not staying with us, hopefully.

File: src/js/relayclient/RelayClient.js (hunk 7 above)
Comment (forshtat) [reply src/js/relayclient/RelayClient.js]: Will do. I think this will need to be done in the scope of StakeManager separation, so didn't make the change here.

File: src/js/relayclient/RelayClient.js (hunk 2 above)
Comment (forshtat) [reply src/js/relayclient/RelayClient.js]: Done

File: src/js/relayclient/RelayClient.js (hunk 8 above)
Comment (forshtat) [reply src/js/relayclient/RelayClient.js]: Done

File: src/js/relayclient/ServerHelper.js (hunk 1 above)
Comment (forshtat) [reply src/js/relayclient/ServerHelper.js]: Can I pull this task to a different PR? This one is already getting bloated and will get in the way of merging with Shahaf

File: test/RelayHubPenalizations.test.js [hunk 9]
```diff
@@ -119,7 +120,7 @@ contract('RelayHub Penalizations', function ([_, relayOwner, relay, otherRelay,
           const txDataSigA = getDataAndSignature(encodeRelayCallEIP155(encodedCallArgs, relayCallArgs), chainId)
           const txDataSigB = getDataAndSignature(encodeRelayCallEIP155(
             encodedCallArgs,
-            Object.assign(relayCallArgs, { gasPrice: 70 }) // only gasPrice may be different
+            Object.assign({}, relayCallArgs, { gasPrice: 70 }) // only gasPrice may be different
```
[the remark this answers is not in the document]
Comment (forshtat) [reply test/RelayHubPenalizations.test.js]: potató-potáto

File: test/server_helper_test.js [hunk 10]
```diff
@@ -146,13 +200,13 @@ contract('ServerHelper', function (accounts) {
       { relay: '2' },
       { relay: '3' },
       { relay: '4', unstakeDelay: 3600 * 24 * 7 }, // dropped out by default, below minDelay
-      { relay: '5', stake: 1e18, transactionFee: 1e5 }, // dropped out by default, below minStake
-      { relay: '6', stake: 3e18, transactionFee: 1e9 },
-      { relay: '7', transactionFee: 1e7 }
+      { relay: '5', stake: 1e18, pctRelayFee: 1e5 }, // dropped out by default, below minStake
```
[the remark this answers is not in the document]
Comment (forshtat) [reply test/server_helper_test.js]: These values are not actually used, they are supposed to test the filtering works; server-helper and its tests need a rewrite badly

File: .eslintrc.js [hunk 11]
```diff
@@ -17,7 +17,7 @@ module.exports = {
       "standard"
     ],
     "parserOptions": {
-        "ecmaVersion": 2018
+        "ecmaVersion": 2020
```
[the remark this answers is not in the document]
Comment (drortirosh) [reply .eslintrc.js]: yes, its a linter. you tell it to use ECMA2020 features.

File: contracts/RelayHub.sol [hunk 12]
```diff
@@ -333,32 +331,32 @@ contract RelayHub is IRelayHub {
...
             // We now verify the legitimacy of the transaction (it must be signed by the sender, and not be replayed),
-            // and that the sponsor will agree to be charged for it.
-            uint256 preconditionCheck;
-            (preconditionCheck, recipientContext) =
+            // and that the paymaster will agree to be charged for it.
+            uint256 canRelayStatus;
+            (canRelayStatus, recipientContext) =
             // TODO: this new RelayRequest is needed because solc doesn't implement calldata to memory conversion yet
             canRelay(
                 GSNTypes.RelayRequest(
```
[the remark this answers is not in the document]
Comment (drortirosh) [reply contracts/RelayHub.sol]: then add a comment. maybe future solc will work, and then we'll know why we put it here.

File: src/js/relayclient/ServerHelper.js (hunk 1 above)
Comment (drortirosh) [reply src/js/relayclient/ServerHelper.js]: what you're saying is "its a baseFee pull request, but a different PR will actually make it use the baseFee"

File: .eslintrc.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (drortirosh) [reply .eslintrc.js]: ... and it actually passes lint without it. please remove.

File: src/js/relayclient/ServerHelper.js (hunk 1 above)
Comment (drortirosh) [reply src/js/relayclient/ServerHelper.js]: then rename it. its "initial support", and in the comment mention that its incomplete.

File: .eslintrc.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (forshtat) [reply .eslintrc.js]: I stopped using them, but `BigInt` and bigint literals (`100n`) are added to eslint at 2020 level.

File: contracts/RelayHub.sol (hunk 12 above)
[the remark this answers is not in the document]
Comment (forshtat) [reply contracts/RelayHub.sol]: Line 343 at your service

File: contracts/samples/KnownUsersSponsor.sol (hunk 4 above)
Comment (forshtat) [reply contracts/samples/KnownUsersSponsor.sol]: Dead code, been dead for a long time, keeps interfering with refactoring and stuff. Not worth keeping. 
Will need a separate sample repository with all of these well-maintained one day.

Comment (drortirosh) [reply contracts/samples/KnownUsersSponsor.sol]: sample code is in the main repository so that refactoring will require maintaining it.
putting it elsewhere means you can keep ignoring maintaining a sample
(yes, we do need to at a test to run this sample, but not having the test right now doesn't mean we need to remove it instead of refactoring it)

Comment (forshtat) [reply contracts/samples/KnownUsersSponsor.sol]: We are currently ignoring the sample 100%.
The way we interact with recipients changed. The APIs changed. This is just some random solidity.
I am not maintaining it just because. The sample is NOT part of the repo. Repo should not contain unrelated code.

File: src/js/relayclient/ServerHelper.js (hunk 1 above)
Comment (forshtat) [reply src/js/relayclient/ServerHelper.js]: Done
